core/src/main/scala/kafka/coordinator/transaction/TransactionStateManager.scala

```diff
@@ -307,28 +306,16 @@ class TransactionStateManager(brokerId: Int,
             && !shuttingDown.get()
             && inReadLock(stateLock) {loadingPartitions.exists { idAndEpoch: TransactionPartitionAndLeaderEpoch =>
               idAndEpoch.txnPartitionId == topicPartition.partition && idAndEpoch.coordinatorEpoch == coordinatorEpoch}}) {
-            val fetchDataInfo = log.read(currOffset, config.transactionLogLoadBufferSize, maxOffset = None,
-              minOneMessage = true, isolationLevel = IsolationLevel.READ_UNCOMMITTED)
-            val memRecords = fetchDataInfo.records match {
-              case records: MemoryRecords => records
-              case fileRecords: FileRecords =>
-                val sizeInBytes = fileRecords.sizeInBytes
-                val bytesNeeded = Math.max(config.transactionLogLoadBufferSize, sizeInBytes)
-
-                // minOneMessage = true in the above log.read means that the buffer may need to be grown to ensure progress can be made
-                if (buffer.capacity < bytesNeeded) {
-                  if (config.transactionLogLoadBufferSize < bytesNeeded)
-                    warn(s"Loaded offsets and group metadata from $topicPartition with buffer larger ($bytesNeeded bytes) than " +
-                      s"configured transaction.state.log.load.buffer.size (${config.transactionLogLoadBufferSize} bytes)")
-
-                  buffer = ByteBuffer.allocate(bytesNeeded)
-                } else {
-                  buffer.clear()
-                }
-                buffer.clear()
-                fileRecords.readInto(buffer, 0)
-                MemoryRecords.readableRecords(buffer)
+
+            val (memRecords, readBuffer) = CoordinatorUtils.readRecords(log, currOffset, buffer,
+              config.transactionLogLoadBufferSize)
+            // Only warn the first time we expand over the configured buffer size
+            if (readBuffer != buffer && readBuffer.capacity > config.transactionLogLoadBufferSize) {
+              warn(s"Loaded transaction metadata from $topicPartition with buffer larger " +
+                s"(${readBuffer.capacity} bytes) than configured ${KafkaConfig.TransactionsLoadBufferSizeProp} " +
+                s"(${config.transactionLogLoadBufferSize} bytes)")
```

Comment:
Are these warnings actually useful? I would say they are not, so I would prefer to remove them. Thoughts @dhruvilshah3 @hachikuji?

Reply:
Hmm yeah, this by itself may not be too useful except when we have corrupted data in the log. I don't think we have a great story around the kind of behavior to expect in this situation, but may be good to leave the log message just so we can trace back to this?

Reply:
Corrupted data will fail the checksum right? That seems more useful than blindly warning due to the size.

Reply:
We can also check if the size is larger than the segment size or something like that.

Reply:
Given we don't checksum the length field, I think the best we could do is to verify if the length falls off the end of the segment like you said. The only annoyance is that making that check at this layer could be tricky given we don't have access to `LogSegment`. But I agree that if we have a way to assert on reading corrupted records, the warning becomes useless at that point.